packages/core/contracts/zodiac/OptimisticOracleModule.sol

```diff
@@ -0,0 +1,197 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.6;
+
+import "@gnosis.pm/zodiac/contracts/core/Module.sol";
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import "../oracle/implementation/Constants.sol";
+import "../oracle/interfaces/FinderInterface.sol";
+import "../oracle/interfaces/SkinnyOptimisticOracleInterface.sol";
+import "../oracle/interfaces/OracleAncillaryInterface.sol";
+
+contract OptimisticOracleModule is Module {
+    event OptimisticOracleModuleDeployed(address indexed owner, address indexed avatar, address target);
+
+    event TransactionsProposed(uint256 indexed proposalId, address indexed proposer, uint256 indexed proposalTime);
+
+    event TransactionExecuted(uint256 indexed proposalId, uint256 indexed transactionIndex);
+
+    event ProposalDeleted(uint256 indexed proposalId);
+
+    FinderInterface public finder;
+    SkinnyOptimisticOracleInterface optimisticOracle;
+
+    IERC20 public collateral;
+    uint256 public bond;
+    string public rules;
+    uint256 public liveness;
+    bytes32 public identifier = "ZODIAC";
+
+    struct Transaction {
```

Comment:
You could save some gas by storing hashes of transactions, rather than the whole transaction object. Similar to what we did for the Reality module.

Reply:
Out of curiosity, how much gas are we talking? The trade-off here is that UMA's dev tools (and voter community) are very familiar with these kinds of structs, since they're how the `Governor` contract works, and less familiar with dealing with transaction hashes, so there would be some extra effort to use transaction hashes, which *might* be worth it if the savings are significant.

Reply:
I think this is something we can estimate at the back of the envelope (and I think doing this regularly is useful!):

We pay 20k gas to write 32 bytes of data assuming that we're not reusing slots we've already used (which is true in this case).
```
    // 96 bytes + length of data (min 128 bytes)
    struct Transaction {
        address to; // 20 bytes, but will use a full 32 byte slot because not next to any smaller data types
        uint256 value; // 32 bytes
        bytes data; // 32 bytes + length of data (min 32 bytes)
        Enum.Operation operation; // 1 byte, could be moved next to the to address so it gets added to that slot
    }

   // 128+ bytes * num_txns + 128+ bytes
   struct Proposal {
        Transaction[] transactions; 128 bytes min * num_txns
        uint256 requestTime; 32 bytes
        bytes ancillaryData; 32 bytes + length of data (min 64 bytes)
        bool status; // 1 byte, but uses a full 32 byte slot
    }
```

This results in a min 256 bytes for 1 transaction, which would cost about 160k gas. This is the minimum. As ancillary data increases and the transaction data grows, this can get much larger.

If we hashed the _entire_ structure, that would reduce all of this down to a single 32 byte slot, which would cost 20k gas, which is about how much a basic ETH transfer costs.

You could also hash each transaction, but keep the outer struct intact. This would help, but wouldn't reduce gas costs nearly as much since each transaction would add about 32k gas.

So I think the gas savings are considerable.

The big disadvantage to doing this is that this would require callers who want to do anything with this data to pass the data in, so it can be verified against the hash. This makes the UX of interacting with the contract more painful, and might mean that users _have_ to rely on external tooling to make use of these contracts.

Reply:
I think a gas efficient version of the module contract would be great in the future but I have data availability concerns. For this initial release, I'm going to stick with putting the full transactions on-chain to ensure that DVM voters can always see them if there is a dispute.

Over time, I think we'll get more tooling and interfaces for interacting with the module and best practices that will make data availability less of a concern. For instance, people will want to ensure that explanations and discussions around proposals are always available. It would make sense to keep the full details of the transactions in the same place.

Then if you had a second version of the module that uses hashed transactions instead of full transactions, and had good reason to believe that the full transactions will almost always be available somewhere else, you might feel comfortable using that version. (This is already the case if the proposal was created and voted on through Snapshot, once this module is integrated there.)

And if the data turns out to be unavailable, the voters would need to reject the proposal. You could also imagine an avatar approving *both* modules and people mainly using the hashed transaction version while keeping the full transaction version around as a back-up.

Reply:
@mrice32 pointed out in a separate thread that we might be able to solve for data availability by emitting the full transaction details in events while only storing the hash on-chain, so I'm backtracking on what I just posted and will give that a shot.

Reply:
I made some updates to the flow. Now, when making a proposal, the contract uses all of the relevant information to encode a `proposalHash`. To execute the proposal, the same data (including transaction data) is entered by the caller and we verify that it generates the same hash that was stored in the contract and approved by the Optimistic Oracle. Then, we can iterate through the transactions in the calldata to execute them one by one.